Proposal to formalize 'language mode' terminology

dempseyatgithub opened the pull request:
Proposal to formalize the term _language mode_ in tool options and Swift PM APIs.

Most discussion occurred in the [first pitch thread](https://forums.swift.org/t/pitch-formalize-swift-language-mode-naming-in-tools-and-api/71733)

The proposal has a [Swift PM implementation](https://github.com/apple/swift-package-manager/pull/7620) which has passed CI tests and has an approval.

dempseyatgithub committed:
Add proposal to formalize 'language mode' terminology

Proposal to formalize the term 'language mode' in tool options and Swift PM APIs.

rjmccall commented:
Okay, two pieces of LSG feedback here about the design:
- Please just immediately deprecate the `swiftLanguageVersion` API from SE-0435 rather than removing it entirely.  There's no harm in having both of these, and we should point the people who've started adopting this at the new API.
- The `swiftLanguageVersions` property and `SwiftVersion` type should be deprecated in the 6.0 tools version rather than obsoleted.  Obsolescence on the initializer is necessary in order to deal with the ambiguity when the argument isn't provided at all, but we should be less demanding when it isn't required.

We'll try to find someone to help you with the driver implementation here.

dempseyatgithub commented:
Thank you for the feedback. I will be making these changes in the next few days.

dempseyatgithub committed:
Incorporate LSG feedback
- Make swiftLanguageVersions property deprecated not obsoleted
- Deprecate Swift setting added in SE-0435 in favor of renamed setting
- Add Acknowledgments section
- Fix typos

dempseyatgithub commented:
I've incorporated the feedback and updated the proposal.

dempseyatgithub committed:
Add deprecation of SwiftVersion typealias
- Improve source compatibility section

rjmccall committed:
Assign SE-0441 to the language mode proposal and put it in review

rjmccall commented:
Do we have any sort of integrated toolchain with all of these changes applied, across both Swift and SPM?

xedin commented:
Toolchain built for SwiftPM changes - https://github.com/swiftlang/swift/pull/73908
